### ISSUE TITLE
nightly `unused_assignments` false positives for `miette::Diagnostic` fields

### ISSUE THREAD

**Issue by adr1anh** [rust-lang/rust#147648]
Starting with rustc 1.92.0-nightly (2300c2aef, 2025-10-12), the compiler’s `unused_assignments` lint now emits warnings in the following snippet which uses `miette` and `thiserror`

### Code

I tried this code:

```rust
#![warn(unused)]

use miette::{Diagnostic, NamedSource, SourceSpan};

#[derive(Debug, thiserror::Error, Diagnostic)]
enum MyError {
    #[error("malformed input")]
    Example {
        #[source_code]
        source_code: NamedSource<String>, // WARNING: value assigned to `source_code` is never read
        #[label]
        span: SourceSpan, // WARNING: value assigned to `span` is never read
    },
}

fn main() {
    let _error = MyError::Example {
        source_code: NamedSource::new("example", "fn main() {}".to_owned()),
        span: SourceSpan::from((0, 5)),
    };
}
```

I expected to see this happen: The code compiles without any error 

Instead, this happened: 

```
warning: value assigned to `source_code` is never read
  --> src/main.rs:10:9
   |
10 |         source_code: NamedSource<String>, // WARNING: value assigned to `source_code` is never read
   |         ^^^^^^^^^^^
   |
   = help: maybe it is overwritten before being read?
note: the lint level is defined here
  --> src/main.rs:1:9
   |
 1 | #![warn(unused)]
   |         ^^^^^^
   = note: `#[warn(unused_assignments)]` implied by `#[warn(unused)]`

warning: value assigned to `span` is never read
  --> src/main.rs:12:9
   |
12 |         span: SourceSpan, // WARNING: value assigned to `span` is never read
   |         ^^^^
   |
   = help: maybe it is overwritten before being read?

warning: `clippy-error` (bin "clippy-error") generated 2 warnings
```

### Version it worked on

It most recently worked on: 
```
rustc 1.92.0-nightly (be0ade2b6 2025-10-11)
binary: rustc
commit-hash: be0ade2b602bdfe37a3cc259fcc79e8624dcba94
commit-date: 2025-10-11
host: aarch64-apple-darwin
release: 1.92.0-nightly
LLVM version: 21.1.2
```

### Version with regression

`rustc --version --verbose`:
```
rustc 1.92.0-nightly (2300c2aef 2025-10-12)
binary: rustc
commit-hash: 2300c2aef7dbc2a7bbbeaa9894d07d459abd9bc6
commit-date: 2025-10-12
host: aarch64-apple-darwin
release: 1.92.0-nightly
LLVM version: 21.1.2
```

### Backtrace

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

**Comment by kpreid** [comment]
I just encountered a similar issue with `displaydoc`. I don't know about what `miette` does, but in `displaydoc`'s case, it is producing the **wrong diagnostic**, `unused_assignments` rather than `unused_variables`. Specifically, given this test program:

```rust
#[derive(displaydoc::Display)]
pub enum Error {
    /// Foo
    Foo { x: String },
}
```

there is an `unused_assignments` warning, but if I copy-paste the expansion into the program unchanged:

```rust
// #[derive(displaydoc::Display)]
pub enum Error {
    /// Foo
    Foo { x: String },
}

#[allow(non_upper_case_globals, unused_attributes, unused_qualifications)]
const _: () = {
    impl ::core::fmt::Display for Error {
        fn fmt(&self, formatter: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
            #[allow(unused_variables)]
            match self {
                Self::Foo { x } => {
                    write!(formatter, "Foo")
                }
            }
        }
    }
};
```

then there is no warning (except for a suppressed `unused_variables`). Thus, something odd is going on here.

**Comment by apiraino** [comment]
Assigning priority (discussion on [Zulip](https://rust-lang.zulipchat.com/#narrow/channel/245100-t-compiler.2Fprioritization.2Falerts/topic/.23147648.20nightly.20.60unused_assignments.60.20false.20positives.20for.20.60m.E2.80.A6/near/544619428)).

@rustbot label -I-prioritize +P-medium -e-needs-mcve

**Comment by Lynnesbian** [comment]
This also happens with [relm4](https://relm4.org):

```rust
use relm4::{prelude::*};

struct Example;

#[relm4::component(pub)]
impl Component for Example {
  type CommandOutput = ();
  type Init = ();
  type Output = ();
  type Input = ();

  view! {
    gtk::ApplicationWindow {
      #[name(my_label)]
      gtk::Label { }
    }
  }

  fn init(_init: Self::Init, root: Self::Root, _sender: ComponentSender<Self>) -> ComponentParts<Self> {
    let model = Example;
    let widgets = view_output!();
    widgets.my_label.set_label(":)");

    ComponentParts { model, widgets }
  }
}

fn main() {
  let app = RelmApp::new("org.example.issue");
  app.run::<Example>(());
}
```

Using the latest beta version of Rust (`1.92.0-beta.3`):

```text
warning: value assigned to `my_label` is never read
  --> src/main.rs:14:11
   |
14 |             #[name(my_label)]
   |                    ^^^^^^^^
   |
   = help: maybe it is overwritten before being read?
   = note: `#[warn(unused_assignments)]` (part of `#[warn(unused)]`) on by default
```

**Comment by lambdageek** [comment]
This is now happening on stable
